### ISSUE TITLE
Discourse should behave like a standard docker image

### ISSUE THREAD

**Issue by philhug** [discourse/discourse_docker#186]
a docker user excepts to be able to run discourse without a special launcher.
to define multiple containers (redis, postgres, discourse,..), links and volumes, docker-compose should be used instead.
https://docs.docker.com/compose/

**Comment by LongLiveCHIEF** [comment]
:+1:   :100: 

I would add, reasons for this being:

- you launcher has a config file that needs to change based on the machine it's located on, which defeats the purpose of using Docker in the first place!
- It's difficult to version control your config file because it doesn't run on any machine the same way (from local to dev to prod) (meaning i have to change things like the "application url" setting for each environment, and that messes with version controlling this file
- your launcher assumes a specific OS (Ubuntu 14), and the whole point of docker is to be OS agnostic
 - your launcher assumes docker executable is in specific path(s) that change from OS to OS, and you wouldn't even need to bother with this using docker the right way, since you know docker is there
- the launcher doesn't benefit from the ability to use docker env variables (such as proxy settings for git), which cause your `bootstrap` to fail in all but the strictest scenarios (aka... using a Digital Ocean server is about the only way to use Discourse). Again... this is the exact thing Docker is meant to solve

I could go on... but I think the basic point comes down to this... you're not really using docker. You just are using base images inside of docker containers, and then adding ruby on top to use base images, instead of using Docker layers to add the customizations you need.

I'd be willing to assist in fixing, if I can get the owner/collaborators to agree that this course of action needs to be taken, since doing it the right way is going to be a "breaking" change for your users already following the `./launcher` procedures.

**Comment by SamSaffron** [comment]
You are going to have to do a proof of concept.

**Comment by LongLiveCHIEF** [comment]
Sounds fun!  I've been doing some tinkering on this the last few days, and what it looks like to me is that this would basically start with taking all of the pups stuff, and working them into dockerfiles for the pups templates, and then using docker-compose to wire it all together.


So far, I've worked through the most common `launcher` commands, and here's how they map out when switching from pups to docker-compose:

| `./launcher` ... | new command (`docker-compose` unless otherwise noted) |
| =========== | ============ |
| bootstrap app | reload/up *deprecated* |
| start app | up |
| stop | stop |
| reload | restart |
| cleanup | rm |

This means the changes for users running the app will look like this:

- bootstrapping and start would happen at the same time, subsequent `up`'s would start or reload configs automatically based on image caches (being changed or not)
- allows for "option" containers (for example a container for SSO vs standard login). this could be huge, because it will automatically give you a "plugin" system for contributors to extend discourse [by contributing compose files](http://docs.docker.com/compose/extends/)
- `rm` automatically removes stopped service containers
- removes the `install docker` and `install git` requirements from the setup
- natively support auto-restart/auto-cleanup for any container inside compose files
- supports environment/custom overrides via environment compose files (so your users can have say a `staging.yml` file and a `production.yml` file to [override or extend the default configuration](http://docs.docker.com/compose/production/))

@SamSaffron if this still sounds interesting, then I'll proceed with working through a POC fork/PR.

**Comment by LongLiveCHIEF** [comment]
One last thought...  If we do this, you could still have a custom launcher script, but it would basically be an interface to executing `docker-compose` commands, but still have a relevant "non-developer friendly" management script.  

I'd suggest changing it though, from `./launcher` to `discourse`. 

I believe taking this course of action would close the following issues:  #184, #174, #169, #146, #140, #124, #79, #17

**Comment by SamSaffron** [comment]
I think the best thing for you to do is open a topic on meta

Explain what issues you are facing with launcher today, what your proposal
is, how you will maintain backwards compat, how will template mixins work,
like the rate limiting one, how the new bootstrapping will work and so on

I don't want this discussion lost on a github issue and I want to make sure
what you are proposing is reasonable before you start working

**Comment by SamSaffron** [comment]
Note keep in mind all our multi machine deployment use launcher, we build
an image with it. Publish to registry and then launch exact same image on
multiple machines using env vars

**Comment by LongLiveCHIEF** [comment]
Here's why I think this is an issues problem, and not a meta discussion. You're using docker, and it's the *only* installation method you support.  The problem is... docker images should run on any Docker machine, without requiring additional installations and unsupported configurations (ex... git). This product does not do that. 

This is an issue (that I'm willing to contribute to), whereby the "install anywhere" capability that Docker adds is not realized, right now, this package only installs in one specific environment.

> Explain what issues you are facing with launcher today, what your proposal
is

That's *exactly* what my previous posts do.

**Comment by SamSaffron** [comment]
This discussion must exist on meta discourse not here

**Comment by LongLiveCHIEF** [comment]
> This discussion must exist on meta discourse not here

This is the purpose github has served the open source community to an exceptional degree, for years. There are excellent reasons for this. Imagine if I had to go to a project specific website for each project I intended to contribute to, and forks/issues/commits couldn't be tracked back to issues.

**Comment by LongLiveCHIEF** [comment]
simply put, if you want me to contribute, I'll do it here. If you make me go somewhere else and introduce un-needed difficulties, then I'll take a pass. I have enough headaches with my day job :-)

**Comment by SamSaffron** [comment]
You are proposing a radical change to the way we distribute Discourse I am
happy to explain, in detail, the reasoning behind the current design, the
implications of a change and the considerations that need to be taken

But this kind of discussion belongs on meta.discourse.org where it can
benefit the rest of the community and be easily searchable

I am not going to spend half an hour explaining my system in some archaic
github ticket

**Comment by philhug** [comment]
@LongLiveCHIEF I'd be happy to see such a PoC and would contribute patches, if the goal is to have 1 image, that is autobuilt and that can be configured with ENV variables and sets up itself on first startup like a typical docker image, I'm in.

**Comment by LongLiveCHIEF** [comment]
@philhug I'll fork and create a milestone with a few issues/requirements that need to be hit before we can submit as a PR. I'll add you as a contributor to the fork.

The issues mostly match the concerns we've all listed above, in addition to the tagged issues, and the requirements to meet mostly match @SamSaffron's comments about build and deployment.

**Comment by LongLiveCHIEF** [comment]
Milestone for tracking efforts/progress towards this issue: https://github.com/HackerHappyHour/discourse_docker/milestones/Dockerize%20discourse_docker

**Comment by SamSaffron** [comment]
@LongLiveCHIEF 

closing this in favor or 

https://meta.discourse.org/t/can-discourse-ship-frequent-docker-images-that-do-not-need-to-be-bootstrapped/33205 

appreciate your desire to help here, please be sure to read that post before starting any work.

**Comment by LongLiveCHIEF** [comment]
:ok_hand:
